Autodiscovery

takeda opened the pull request:
Adding autodiscovery support as described here: https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.AddingToYourClientLibrary.html

Fixes: https://github.com/emcache/emcache/issues/97

pfreixes commented on emcache/cluster.py:
```diff
@@ -221,6 +252,60 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+            except ValueError:
+                new_nodes.append(Node(MemcachedHostAddress(host, port), *self._new_node_options))
+
+        self._healthy_nodes = new_nodes
```
Not sure if we should go and add all of them as healthy nodes, for the ones that already existed in previous versions and they were marked as unhealthy IMO we should keep them as unhealthy and only having them recovered back if by them selves mark them as healthy again.

Another option, is that instead of rebuilding the whole list we just add the new ones and remove the no longer available, while we keep the ones that were already there but without adding them again. This should allow us to not have to change any healthy status

pfreixes reviewed (changes requested):
Great PR, thank you for the contribution!!

A couple of general comments:

- Could you add the proper documentation? In the README.MD under its own section and also as part of the [docs](https://github.com/emcache/emcache/blob/master/docs/index.rst)
- How can we improve confidence on the changes regarding testing? My main concern - not sure if there is a good or even a solution - is with the lack of acceptance tests that allow us to check that this solution works

takeda commented on emcache/client.py:
```diff
@@ -626,6 +630,7 @@ async def create_client(
     ssl: bool = DEFAULT_SSL,
     ssl_verify: bool = DEFAULT_SSL_VERIFY,
     ssl_extra_ca: Optional[str] = None,
+    autodiscovery: Optional[float] = None,
```
It specifies how frequently to check for changes. The `float` is in seconds. As the point is to update the node list if there's a change in the cluster.

takeda replied:
Regarding the second paragraph. It is true, that AWS provides it but I see it implemented in GCP as well so it starting to be a semi standard: https://cloud.google.com/memorystore/docs/memcached/about-auto-discovery

takeda replied:
@pfreixes I made the setting into boolean and also added poll interval (60s) and timeout (5s). I don't think there is a good reason to call it aws_autodiscovery, since looks like Google Cloud also adopted it, do you agree?

takeda commented on emcache/cluster.py:
```diff
@@ -71,38 +80,52 @@ def __init__(
         self._cluster_events = cluster_events
         self._purge_unhealthy_nodes = purge_unhealthy_nodes
+        self._autodiscovery = autodiscovery
+        self._autodiscover_version = 0
```
I was thinking about it, but it looks like it always start with 1. How about using -1? This would avoid having additional if statement?

takeda replied:
fixed, but I used -1, I noticed modified by AWS Java client used -1 for invalid, so that value is very unlikely to be ever used

takeda commented on emcache/cluster.py:
```diff
@@ -221,6 +252,60 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+                break
+
+            except Exception:
+                logger.exception(f"Autodiscover raised an exception, continuing processing ")
```
This was based on `_events_dispatcher()`'s code

takeda commented on emcache/cluster.py:
```diff
@@ -71,38 +80,52 @@ def __init__(
         self._closed = False
+
+        if self._autodiscovery:
+            asyncio.ensure_future(self.autodiscover())
```
Need to think about how to do #1, as for #2 I don't have a good idea how I would accomplish it. The events task executes whenever a new event is being added, while this task supposed to execute periodically (e.g. every 60 seconds). It makes more sense to me that they are separate tasks.

takeda replied:
@pfreixes I did #1 as for #2 I don't have idea how I could do it with just one task

takeda commented on emcache/cluster.py:
```diff
@@ -221,6 +252,60 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+            raise RuntimeError("Emcache client is closed")
+
+        node = random.choice(self._original_nodes)
+        async with OpTimeout(self._timeout, self._loop):
```
Added a custom timeout setting.

takeda commented on emcache/cluster.py:
```diff
@@ -221,6 +252,60 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+                return await connection.autodiscovery()
+
+    async def autodiscover(self) -> None:
+        autodiscovery, version, nodes = await self._get_autodiscovered_nodes()
```
fixed

takeda commented on emcache/cluster.py:
```diff
@@ -221,6 +252,60 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+        new_nodes: list[Node] = []
+        for host, ip, port in nodes:
+            try:
+                new_nodes.append(self.node(MemcachedHostAddress(host, port)))
```
This is no longer relevant as the code was modified to preserve the health status of the nodes. The try-catch was no longer needed.

takeda commented on emcache/protocol.py:
```diff
@@ -22,6 +23,61 @@
 NOT_FOUND = b"NOT_FOUND"
 
 
+class AutoDiscoveryParser:
```
done

takeda commented on emcache/cluster.py:
```diff
@@ -221,6 +252,60 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+                break
+
+            try:
+                await asyncio.sleep(self._autodiscovery)
```
the sleep time was exposed as `autodiscovery_polling_interval`.

pfreixes commented on emcache/cluster.py:
```diff
@@ -221,6 +254,80 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+    async def _autodiscovery_monitor(self) -> None:
+        logger.debug("Autodiscovery task started")
+        while True:
+            if not self._autodiscovery:
```
Is this really necessary? so if `_autodiscovery_monitor` was executed was because `self._autodiscovery` was configured as true, right? if so seems to me that this is not really necessary.

takeda replied:
You're right, I will remove it.

pfreixes commented on emcache/cluster.py:
```diff
@@ -221,6 +254,80 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+            async with node.connection() as connection:
+                return await connection.autodiscovery()
+
+    async def autodiscover(self) -> bool:
```
Its been a while since I worked with Python and Asyncio, but could we have an instance variable for storing the result of the first success autodiscovery execution? With that we could avoid having to run multiple `autodiscovery` commands [here](https://github.com/emcache/emcache/pull/98/files#diff-d5918e537eadb7095bd9b366b7a19681b87d8d0f87683d5d4007a6fa4d5e861cR720) and [here](https://github.com/emcache/emcache/pull/98/files#diff-671bcd2cafefff8bcd836e45326e78a939d6c6276631c3976a8b5967fd27a54aR282)

IIRC we could do something like:

```python
self._first_autodiscovery_done = loop.create_future()
```

And then we could set the value when the first `autodiscovery` function would be successfully executed, something like:

```python
if not self._first_autodiscovery_done.done():
  self._first_autodiscovery_done.set_result(true)
```

And then later [here](https://github.com/emcache/emcache/pull/98/files#diff-d5918e537eadb7095bd9b366b7a19681b87d8d0f87683d5d4007a6fa4d5e861cR720) we could do something like:

```python
if autodiscovery:
  await asyncio.wait_for(client._cluster._first_autodiscovery_done, DEFAULT_STARTUP_WAIT_AUTODISCOVERY)
```

Thoughts?

takeda replied:
I updated the code to work that way. Yeah, it's better.

pfreixes reviewed (commented):
Thanks for the updates, added some more comments!

takeda commented on docs/client.rst:
```diff
@@ -40,6 +40,9 @@ When no other parameters are provided, the following keyword arguments are confi
 - **ssl** By default False. If enabled will make the connection using SSL/TLS protocol.
 - **ssl_verify** By default True. If enabled, will verify if the server certificate is trustable.
 - **ssl_extra_ca** By default `None`. If provided will used as an extr CA file used during the certificate verify. Use together `ssl_verify` when needed.
+- **autodiscovery** By default, `False`. If enabled the client will automatically call `config get cluster` and update node list.
```
It is the same thing, here is more details: https://cloud.google.com/memorystore/docs/memcached/use-auto-discovery

They also use `config get cluster` and the output is same as described in AWS document:

```
CONFIG cluster 0 [length-of-payload-in-next-two-lines]
[integer]
[node1-ip]|[node1-ip]|[node1-port][node2-ip]|[node2-ip]|[node2-port]
\r\n
END\r\n
```

Though if you still think it should be prefixed with "aws_", I can update that.

pfreixes replied:
You are right, protocol is the same ... but the way that the command is issued seems different right? so in the case of Google it uses an ad-hoc discovery endpoint, while for AWS you can use any cluster node or just use the domain name provided by AWS. If this is true the way that the discovery is done would differ a bit, so we would need to differentiate this in some way.

In any case, unless Im missing something, we do not have any way of testing this, so Im expecting that you will test the implementation with just AWS, is that correct? if so the options that we have are the following

1 - We keep the generic `autodiscovery` parameters name, but we make sure that documentation clarifies that this is works Today for AWS
2 - We rename the parameters using `aws_blabla`
3 - We create a new ad-hoc factory called `create_aws_discovery_client` which would accept al current parameters plus the ones related to autodiscovery.

IMO 3 seems the cleanest way and less error prone, thought we should maintain a bit more code but IMO would pay off.

Thoughts?

takeda replied:
That part also is the same. AWS has a similar "main" endpoint. Which randomly resolves to one of memcache nodes. The instruction is just how to query their API to obtain it. The DNS name supposed to be static for the entire life of the cluster. This supposed to be the initial endpoint to bootstrap the process.

Yes, the testing is a pain, as this is non standard and I'm trying it out whenever I'm pushing updates to the PR. My code will ultimately just run in AWS so if long term issues show up, I'll make PR to address them. The issue right now that I'm having is that the changes aren't part of the package, so it is a bit more complex and still thinking how I can use unrelased version. Perhaps for now will include it in the app.

I was originally thinking about doing something like #3 and have that custom client in my own code (and not needing a PR), but the difficulty is that this supposed to also periodically check and adapt as cluster changes.

I also think that GCP implementation is identical to AWS, I tried to look for Azure, but am not having much luck. It looks though like they just have their own proprietary solution. And anyone who still wants to use memcached they need to spin up VMs with it and figure out other details. So I think \#1 is more appropriate.

pfreixes replied:
Oks lets go for 1, lets make clear in the documentation when this can be used.

takeda replied:
updated

takeda commented:
@pfreixes I will look into adding unit tests for my code tomorrow

takeda commented:
@pfreixes added unit tests, can you please review again, thx

pfreixes commented on emcache/cluster.py:
```diff
@@ -221,6 +257,82 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+            async with node.connection() as connection:
+                return await connection.autodiscovery()
+
+    async def autodiscover(self) -> bool:
```
Hey Im currently on vacations, Im quite limitted for making reviews, so be patience :)

In any case this function should have proper coverage,  unless it has already, for checking the different branches i.e new nodes added, nodes removed, ...

takeda replied:
I added unit tests and that fuction is now 100% covered.

pfreixes commented on docs/client.rst:
```diff
@@ -40,6 +40,10 @@ When no other parameters are provided, the following keyword arguments are confi
 - **ssl** By default False. If enabled will make the connection using SSL/TLS protocol.
 - **ssl_verify** By default True. If enabled, will verify if the server certificate is trustable.
 - **ssl_extra_ca** By default `None`. If provided will used as an extr CA file used during the certificate verify. Use together `ssl_verify` when needed.
+- **autodiscovery** By default, `False`. If enabled the client will periodically call `config get cluster` and update node list according to its output. Note, this is not a standard feature of memcached, but is necessary to
```
We would need to update the Readme for telling that autodiscovery for AWS is supported, also in the docs when how clients are created would be interesting to describe this functionallity.

takeda replied:
Updated README.rst, could you clarify the second part?

pfreixes replied:
there is this documentation that would be nice to update too https://github.com/emcache/emcache/blob/master/docs/client.rst

takeda replied:
This was editing that file, but I also added an example of calling it.

takeda commented:
Note I also made few additional improvements based on what I've experienced from real usage.

pfreixes commented:
Ill make another review next days, meanwhile lets run the tests and check that all of them pass

takeda commented:
BTW: Not sure if this failure is because of it, will check it locally later, but I noticed that flake8 complains at the formatting black produces. Is it ok to make changes to it?

takeda commented:
I fixed issues with flake8 (I updated it to the latest version and added W503, E203) to ignore as they conflicted with black: https://black.readthedocs.io/en/stable/faq.html#why-are-flake8-s-e203-and-w503-violated

pfreixes commented on emcache/cluster.py:
```diff
@@ -221,6 +273,93 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+
+        # Close unused nodes
+        for node in closable:
+            await node.close()
```
Would make sense to move this to the very end of function, until `self._build_rdz_nodes()` is not executed commands issued might use invalid nodes.

takeda replied:
done

pfreixes commented on emcache/cluster.py:
```diff
@@ -221,6 +273,93 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+        # Updating node list with preservation of their status
+        new_nodes_set = {MemcachedHostAddress(host, port) for host, ip, port in nodes}
+        old_nodes_set = {node.memcached_host_address for node in self.nodes}
+        protected = {node.memcached_host_address for node in self._discovery_nodes}
```
Is this something really necessary? I would expect - please correct me if Im wrong - that AWS provides you a generic hostname for making the discovery which could resolve to any host of the cluster, while the response returned back from the command for making the discovery would return exclusively hostnames that can resolve to the specific node hosts.

If this was true, I would expect that we do not need to make this protection.

A related conversation, is how the usage of a connection pool can impact the discovery? Remember that in the end when we instantiate a Node we are creating a connection pool where here is 1 or multiple connections already opened. Worst case scenario could be tens of them. If one of the nodes of the cluster is removed and we had the back luck of making the resolution to that node, the connection pool could hold invalid TCP connections which would imply that the autodiscovery command would fail. True that eventually all TCP connections would get recycled and eventually new DNS resolutions and new TCP connections will be created.

Wondering if we should be an step ahead on this scenario, we should double check this, and avoid this situation by avoiding the connection pool usage or just use another strategy like making the discovery using all healthy nodes until one succeed.

Thoughts?

takeda replied:
Actually this was a problem when I placed that generic hostname in `self._healthy_nodes` as this is no longer true, this code might indeed not be neeeded anymore.

takeda replied:
removed that code

takeda commented on docs/client.rst:
```diff
@@ -54,6 +58,18 @@ Example of a client creation that would not purge unused connections
+
+    client = await emcache.create_client(
+        [
+            emcache.MemcachedHostAddress('mycluster.fnjyzo.cfg.use1.cache.amazonaws.com', 11211)
```
The address I took from the AWS documentation, so it should be ok to use.

takeda commented on emcache/cluster.py:
```diff
@@ -71,42 +82,75 @@ def __init__(
+                    ssl_extra_ca,
+                )
+                for memcached_host_address in memcached_hosts_address
+            ]
```
Originally I used copy of `self._healthy_nodes` and created `self._original_nodes` and used the later one for discovery, but kept the seed node in `self._healthy_nodes`. The biggest reason is my fear of getting into situation where somehow all nodes were marked as unhealthy and I don't have anything to look up.

Also the configuration hostname `mycluster.fnjyzo.cfg.use1.cache.amazonaws.com` looks very different than the original hostnames for example `0001.mycluster.fnjyzo.use1.cache.amazonaws.com`, `0002.mycluster.fnjyzo.use1.cache.amazonaws.com`. It also behaves differently. The configuration hostname has very short TTL and rotates over different nodes.

When I would add it to `self._healthy_nodes` it would screw up the hashing and make one node be used more than the others. The issue also got bigger when I increased minimum connections, as then I noticed I was making many connections to that seed node, so it was clear this needs to be treated differently, so this and how it affected hashing was the main reason for the change.

pfreixes replied:
Good point. thanks for the explanation!

takeda commented on README.rst:
```diff
@@ -18,6 +18,7 @@ on the experience of other Memcached clients, providing the following main chara
 - Support for many Memcached hosts, distributing traffic around them by using the `Rendezvous hashing <https://emcache.readthedocs.io/en/latest/advanced_topics.html#hashing-algorithm>`_ algorithm.
 - Support for different commands and different flag behaviors like ``noreply``, ``exptime`` or ``flags``.
 - Support for SSL/TLS protocol.
+- Support for `AWS Autodiscovery <https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.html>`_, should also work with GCP memcached cluster.
```
updated

takeda commented on emcache/cluster.py:
```diff
@@ -221,6 +273,93 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+                if await self.autodiscover():
+                    await asyncio.sleep(self._autodiscovery_poll_interval)
+                else:
+                    await asyncio.sleep(AUTODISCOVERY_POLL_INTERVAL_FAILURE)
```
updated

takeda commented on tests/unit/test_cluster.py:
```diff
@@ -132,22 +177,323 @@ async def test_node_initialization(self, mocker, node1, node2, memcached_host_ad
+            False,
+            False,
+            None,
+            False,
```
added comments

pfreixes reviewed (changes requested):
We are "almost" there let's try to fix all comments that Ive added.

Will be still an open discussion about if we should treat the discovery endpoint in a way that it does not make usage of the connection pool and it makes a new TCP connection all the time for also having a chance to gather new IP addresses during DNS resolution, Im happy to have this discussion in some way "separately" but I would be inclined to have this discussion solved before we merge the PR or we make a new release of emcache.

takeda commented on emcache/default_values.py:
```diff
@@ -8,3 +8,6 @@
 DEFAULT_AUTOBATCHING_ENABLED = False
 DEFAULT_SSL = False
 DEFAULT_SSL_VERIFY = True
+DEFAULT_AUTODISCOVERY_POLL_INTERVAL = 60.0
```
At least on AWS changing configuration is surprisingly slow, it takes maybe 10 minutes to add new nodes (not sure what it does during that time).

AWS uses 60s in their fork of java client: https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java/blob/8fffadc84c1b7fe5eff2a163cc04e22ea5902b2d/src/main/java/net/spy/memcached/ConfigurationPoller.java#L38

pfreixes replied:
nice, lets keep it to 60s

takeda commented on emcache/cluster.py:
```diff
@@ -221,6 +272,90 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+            autodiscovery, version, nodes = await self._get_autodiscovered_nodes()
+        except asyncio.TimeoutError as e:
+            logger.error("Got timeout when checking cluster configuration: %r", e)
+            return False
```
I made changes as requested. I never in the past used `create_future` and `wait_for` so I'm hoping I'm using them correctly.

takeda commented on emcache/default_values.py:
```diff
@@ -8,3 +8,6 @@
 DEFAULT_SSL_VERIFY = True
+DEFAULT_AUTODISCOVERY_POLL_INTERVAL = 60.0
+DEFAULT_AUTODISCOVERY_TIMEOUT = 5.0
+DEFAULT_STARTUP_WAIT_AUTODISCOVERY = 120.0
```
updated to 10s

pfreixes commented on emcache/cluster.py:
```diff
@@ -71,42 +81,75 @@ def __init__(
+                Node(
+                    memcached_host_address,
+                    1,
+                    1,
```
I did not see this, this definitely will help us to mitigate any issue with having connections opened to nodes that were removed from the cluster but they were used for making the discovery, so worst case scenario we would have one discovery cycle filing and having the TCP recycled, next attempt should work since a new DNS query should be issued which should point to a new IP.
